### ISSUE TITLE
Disable placeholder when dragging item outside tree

### ISSUE THREAD

**Issue by sberube** [angular-ui-tree/angular-ui-tree#464]
Is there a way to disable the placeholder element when dragging the item outside a tree? If you look at the video, what I want it's when I drop the item outside tree2 I want it to go back to tree1 (rollback the drag'n'drop). Is there any way way to do this or do I need to modify the plugin. If this is the case can someone help me with this?

http://1drv.ms/1Hk5DeS

Thanks!

**Comment by AndrewIsh** [comment]
The way I did this was to catch the dropping of an element and check it's within the model that I want it, and if not, cancel the drop. So:
```
        $scope.listOptions = {
            beforeDrop: function(e) {
                // Check the new parent is a "category" model
                if (!e.dest.nodesScope.$parent.category) {
                    // If not, cancel the drop
                    e.source.nodeScope.$$apply = false;
                }
            }
        };
```

**Comment by sberube** [comment]
@AndrewIsh $parent.category is always undefined. Even when I drag the node inside the tree. Is it normal? Is it something you added?

**Comment by AndrewIsh** [comment]
@sberube Sorry, I wasn't clear. "category" is just the name of the model object I want to restrict dropping to, i.e. if I'm trying to drop and the model object outside my "category" model, I'm prevented from doing so.

So, in your case, whatever you're iterating on (e.g. ng-repeat="node in myModel") is what you'd replace "categories" with. i.e (if your model is called myModel):
```
        $scope.listOptions = {
            beforeDrop: function(e) {
                // Check the new parent is a "myModel" model
                if (!e.dest.nodesScope.$parent.myModel) {
                    // If not, cancel the drop
                    e.source.nodeScope.$$apply = false;
                }
            }
        };
```

**Comment by sberube** [comment]
OK! We managed to cancel the dropping of the node by using the following statement :
```e.source.nodeScope.$$apply = false;```

It will be nice if the documentation can specify that by specifying this line the event can be cancelled.

**Comment by AndrewIsh** [comment]
I have since learnt that it seems the *correct* way of doing this is to make use of the 'accept' method in $callbacks. This should return true or false depending on whether the element currently being hovered over should accept the element. Take a look at the [code near the bottom of the groups example](https://jimliu.github.io/angular-ui-tree/js/groups.js)

**Comment by andreaslarssen** [comment]
I'm looking at the same issue atm. I can't seem to find a way to do this, the ui-tree scope doesn't have a beforeDropped callback, which I feel would be natural. The accept callback fires when a node enters another tree, and when it leaves. Also, it's not really stable; it only fires when you move nodes at a slow pace. 

Am I wrong, @AndrewIsh if I read that the code you're referring to only disable dragging of nodes between trees?

**Comment by AndrewIsh** [comment]
@andreaslarssen The code I'm referring to is the 'accept' callback. What is it you're trying to achieve?

**Comment by andreaslarssen** [comment]
@AndrewIsh: When dragging a node from tree 1 to tree 2, but dropping it outside tree 2 (and 1), the node should go back to tree 1, not 2 like it is by default. It attaches itself to the tree it last was hovered over. I think that's what @sberube also meant by the first post?

**Comment by tpederson** [comment]
There is not a good way I know of to present a confirmation to the user, and either complete the drop or "revert" the drop.

Using this technique was problematic for me:

`e.source.nodeScope.$$apply = false`

This had the effect of canceling the drop, but it also made the dragged node un-draggable ever again.

**Comment by andreaslarssen** [comment]
@tpederson: No, there's no good way to present a confirmation to the user either, which is something I would like to see added (I use this lib to create a dragdrop-to-delete function).

This issue is for nodes automatically dropping to the last hovered tree. Seems like @ppuleo is on to something, though.

**Comment by Voles** [comment]
Closed, this has been fixed with #550 and will be available in the next release.
